### PR TITLE
[GlobalOpt] Fix global SRA incorrect alignment on some elements

### LINKED ISSUE #115282: [GlobalOpt] Reduction of global aggregate with alignment constraint unnecessarily propagates alignment to some elements
When GlobalOpt performs SRA on a global aggregate with an alignment constraint, it seems to propagate the alignment to _some_ of the elements haphazardly. For example:

```
alignas(32) static int params[8];
```

On a plain 32 bit architecture, the array elements respectively have alignment: 32, 4, 8, 4, 16, 4, 8, 4
GlobalOpt SRA creates individual global variables for the elements with respective alignments: 32, 32, 8, 32, 16, 32, 8, 32

I don't see any benefit to increase the alignments elements 1, 3, 5, 7 from 4 to 32. This is wasteful because the new elements cannot be mapped consecutively in memory anymore.

Godbolt demo: https://godbolt.org/z/WjbPjY9Kh

### PR DESCRIPTION
**Pull request by brunodf-gf**
The logic had a flaw where the alignment from the original aggregate is unintentionally retained for elements when the calculated known alignment is not higher than the element's ABI type alignment.

Fixes #115282

### CODE REVIEW

**Comment by brunodf-snps** [comment]
I think the bug existed since commit 4796b4ae7bccc7b75a26b666fbd554c49b274404.

The updated expected alignments in Transforms/GlobalOpt/globalsra-align.ll revert the changes from that commit, with the exception of `@a.6` (previously `@a.2.2`) which before that commit had alignment 8, and since alignment 16. This is retained, but seems correct to me. 

The tests DebugInfo/X86/global-sra-struct-fit-segment.ll and DebugInfo/X86/global-sra-struct-part-overlap-segment.ll are newer than the aforementioned commit.

CC @nikic

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp [hunk 1]
```diff
@@ -575,15 +575,16 @@ static GlobalVariable *SRAGlobal(GlobalVariable *GV, const DataLayout &DL) {
...
 
     // Calculate the known alignment of the field.  If the original aggregate
-    // had 256 byte alignment for example, something might depend on that:
+    // had 256 byte alignment for example, then the element at a given offset
+    // may also have a known alignment, and something might depend on that:
     // propagate info to each field.
     Align NewAlign = commonAlignment(StartAlignment, OffsetForTy);
-    if (NewAlign > DL.getABITypeAlign(Ty))
-      NGV->setAlignment(NewAlign);
+    NGV->setAlignment(std::max(NewAlign, DL.getABITypeAlign(Ty)));
```
Comment (nikic) [inline llvm/lib/Transforms/IPO/GlobalOpt.cpp:587]: Are there any further test changes if we just drop the getABITypeAlign part of this completely?

I see that the later code does a getOrEnforceKnownAlignment, which will raise the global alignment to the preferred alignment anyway, so I suspect that the check for ABI alignment is just entirely unnecessary here.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/IPO/GlobalOpt.cpp (hunk 1 above)
Comment (brunodf-snps) [reply llvm/lib/Transforms/IPO/GlobalOpt.cpp]: Indeed, in the tests leaving out the ABI type align part makes no difference.

**Comment by brunodf-snps** [comment]
> LGTM

Thanks. I still added a reduced test case with the example from the bug report. I suppose this should be pre-committed? In that case, you can commit 0a59028 first, and I'll rebase to remove it from this pull request.

(Unless you think the effect is sufficiently clear from the other test cases and a dedicated test case is overkill; then I'll remove it again.)

**Pull request merged**